video/audio_channel.h

```diff
@@ -28,26 +30,27 @@ class AudioChannel {
 		void		setVolumeEnvelope(std::unique_ptr<VolumeEnvelope> envelope);
 		void		setFrequencyEnvelope(std::unique_ptr<FrequencyEnvelope> envelope);
 		void		setSampleRate(uint16_t sampleRate);
-		WaveformGenerator * getWaveform() { return this->_waveform.get(); }
+		std::shared_ptr<fabgl::WaveformGenerator> getWaveform() { return *(_waveform.get()); }
 		void		attachSoundGenerator();
 		void		seekTo(uint32_t position);
 		void		loop();
 		uint8_t		channel() { return _channel; }
 	private:
-		std::unique_ptr<fabgl::WaveformGenerator>	getSampleWaveform(uint16_t sampleId, std::shared_ptr<AudioChannel> channelRef);
+		std::shared_ptr<fabgl::WaveformGenerator>	getSampleWaveform(uint16_t sampleId, std::shared_ptr<AudioChannel> channelRef);
 		void		waitForAbort();
 		uint8_t		getVolume(uint32_t elapsed);
 		uint16_t	getFrequency(uint32_t elapsed);
 		bool		isReleasing(uint32_t elapsed);
 		bool		isFinished(uint32_t elapsed);
-		AudioState	_state;
-		uint8_t		_channel;
-		uint8_t		_volume;
-		uint16_t	_frequency;
-		int32_t		_duration;
-		uint32_t	_startTime;
-		uint8_t		_waveformType;
-		std::unique_ptr<WaveformGenerator>	_waveform;
+		std::atomic<AudioState>				_state;
+		std::atomic<std::uint8_t>			_channel;
+		std::atomic<std::uint8_t>			_volume;
+		std::atomic<std::uint16_t>			_frequency;
+		std::atomic<std::int32_t>			_duration;
+		// std::atomic<std::uint32_t>		_startTime;
+		std::atomic<std::uint64_t>			_startTime;
+		std::atomic<std::uint8_t>			_waveformType;
+		std::shared_ptr<std::shared_ptr<WaveformGenerator>>	_waveform;
```

Comment:
in the latest/last commit on this branch I decided to try using a pointer-to-a-pointer pattern for the waveform.  the idea being that using an underlying variant of vdp-gl that used smart pointers, this would allow the underlying audio system to retain the waveform object when the agon-vdp layer was done with it

this didn't really help - there were still crashes occurring.

this branch was then abandoned

instead mutexes were added around changing the waveform as well as a different approach inside vdp-gl taken using mutexes around attach/detach.  this is the approach used in #118 which appears to be much more reliable and _seems_ to mostly work